evaluations/src/topk/mod.rs

```diff
@@ -901,42 +921,310 @@ fn check_global_stopping(
+/// Paginates through the entire dataset to fetch all datapoint IDs, respecting
+/// `max_datapoints` if set. The IDs are shuffled for randomized evaluation order.
 #[expect(dead_code)]
 async fn fetch_datapoint_ids_step(
-    _function_name: String,
-    _dataset_name: String,
-    _max_datapoints: Option<usize>,
-    _state: TopKTaskState,
-) {
-    // Implementation will:
-    // 1. Call list_datapoints() to get datapoint IDs from ClickHouse
-    // 2. Shuffle the IDs for randomized evaluation order
-    // 3. Return the shuffled IDs (checkpointed by durable framework)
+    params: FetchDatapointIdsParams,
+    state: TopKTaskState,
+) -> anyhow::Result<Vec<Uuid>> {
+    const PAGE_SIZE: u32 = 1000;
+
+    let mut all_ids: Vec<Uuid> = Vec::new();
+    let mut offset: u32 = 0;
+    let max_to_fetch = params.max_datapoints.map(|n| n as u32);
+
+    loop {
+        // Calculate how many to fetch in this page
+        let limit = match max_to_fetch {
+            Some(max) => {
+                let remaining = max.saturating_sub(offset);
+                if remaining == 0 {
+                    break;
+                }
+                PAGE_SIZE.min(remaining)
+            }
+            None => PAGE_SIZE,
+        };
+
+        let request = ListDatapointsRequest {
+            function_name: Some(params.function_name.clone()),
+            limit: Some(limit),
+            offset: Some(offset),
+            ..Default::default()
+        };
+
+        let response = list_datapoints(
+            &state.clients.clickhouse_client,
+            params.dataset_name.clone(),
+            request,
+        )
+        .await?;
+
+        let page_ids: Vec<Uuid> = response.datapoints.iter().map(|d| d.id()).collect();
+        let page_size = page_ids.len() as u32;
+
+        all_ids.extend(page_ids);
+
+        // Stop if we got fewer than requested (end of dataset)
+        if page_size < limit {
+            break;
+        }
+
+        offset += page_size;
+    }
+
+    let mut ids = all_ids;
+
+    // Shuffle for randomization
+    use rand::seq::SliceRandom;
+    let mut rng = rand::rng();
+    ids.shuffle(&mut rng);
+
+    Ok(ids)
 }
 
-/// Durable step function to process a batch and update progress state.
+/// Extracts the variant name from an `EvaluationVariant`.
+///
+/// For named variants, returns the name directly. For dynamic variants (which don't
+/// have persistent names), returns a placeholder string "dynamic_variant".
+#[allow(dead_code, reason = "used by process_batch_step skeleton")]
+#[allow(
+    clippy::allow_attributes,
+    reason = "expect(dead_code) warns because item is used"
+)]
+fn get_variant_name(variant: &EvaluationVariant) -> String {
+    match variant {
+        EvaluationVariant::Name(name) => name.clone(),
+        // Dynamic variants don't have names, use a placeholder
+        EvaluationVariant::Info(_) => "dynamic_variant".to_string(),
```

Comment:
can you open an issue and assign it to @GabrielBianconi and me to figure out what we want the behavior to be for names for dynamic variants. We should use tensorzero::dynamic here for now.